### PR TITLE
IImage base interface for images

### LINKED ISSUE #909: More flexible image support
Currently all images used anywhere in Pyface or TraitsUI need to be `ImageResource` instances.  This is a limited use-case intended for images shipped with an application and stored as files.

There are frequent cases where a more generic image is to be displayed. If the image comes from a file, then the `ImageResource` infrastructure can be made to work by supplying an absolute path to the image, but for images in memory or from remote sources (eg. a URL) there is no way to display them other than saving out to a temporary file and then reading it back in.

For complex use-cases (eg. image editors) Enable/Chaco is often the right solution, but there are many use-cases where simple display of in-memory or url-based images is desirable.

A possible approach to this is:

- [ ] add an `IImage` interface which includes the `create_bitmap` and `create_icon` methods from `IImageResource`, but makes no assumptions about the location or storage of the data
- [ ] have `IImageResource` be a sub-interface of `IImage`
- [ ] change the `Image` trait to accept any `IImage` rather than any `IImageResource`

Everything up to here is effectively backwards compatible, as it is just injecting a new base interface.

- [ ] create other interfaces and concrete implementations for other types of images, such as `IImageBuffer` for an image stored in a memory buffer, or `IPILImage` for PIL/Pillow images, possibly `IURLImage` for a remote image, etc.

This might cause some problems in the rare case where the code is asking for the absolute path of an `IImageResource` (eg. in the About dialog, where it is generating HTML) but these can either be fixed or the API changed to be more restrictive if needed - worst case we might need a way to flag that the `Image` trait only accepts a particular sub-interface.

ETA: The `convert_bitmap` function also needs to be dealt with - it provides a cache for toolkit bitmaps, but that doesn't make sense for highly dynamic images, particularly if they are sharing a buffer somewhere.  It is used by TraitsUI's image editor, for example.  The simplest modification is to only do caching for `IImageResource` instances, and otherwise just return the bitmap and let the class do any caching needed.

Edit once more: `convert_bitmap` doesn't take into account image sizes; code should use an `ImageCache` instead, or we should re-think image caching.  This is a separate issue.

### PR DESCRIPTION
**Pull request by corranwebster**
This is the minimal functionality to add a new base `IImage` interface underneath `IImageResource` to potentally allow dynamic image types (eg. PIL/pillow images, numpy arrays, kiva GCs, SVG images) to be more easily used with Pyface and TraitsUI.

It implements additional concrete `IImage` class that wrap numpy arrays.

It also includes updates to the `Image` trait so that it accepts `IImage` instances as well as `IImageResource` and strings.

This fixes #909.  It should be backward compatible except as noted there.

### CODE REVIEW

**Comment by rahulporuri** [comment]
@corranwebster is this PR ready for review?

**Comment by corranwebster** [comment]
Yes - I have some extra implementation classes that are on my personal laptop, but they can go into a separate PR.

As of the last commit, tests are working and should be clean.

File: pyface/array_image.py
Comment (rahulporuri) [inline pyface/array_image.py:1]: why add these modules to the top level `pyface` namespace when the more relevant `pyface.image` subpackage exists?

File: pyface/i_image_resource.py
Comment (rahulporuri) [inline pyface/i_image_resource.py:67]: Is there any useful information in this fixme that we want to retain? _i don't think so IIUC_

File: pyface/ui/qt4/util/image_helpers.py
Comment (rahulporuri) [inline pyface/ui/qt4/util/image_helpers.py:25]:
```suggestion
class ScaleMode(Enum):
```

File: pyface/ui/qt4/util/image_helpers.py
Comment (rahulporuri) [inline pyface/ui/qt4/util/image_helpers.py:30]:
```suggestion
class AspectRatio(Enum):
```

File: pyface/ui/wx/util/image_helpers.py
Comment (rahulporuri) [inline pyface/ui/wx/util/image_helpers.py:23]:
```suggestion
class ScaleMode(Enum):
```

File: pyface/ui/wx/util/image_helpers.py
Comment (rahulporuri) [inline pyface/ui/wx/util/image_helpers.py:28]:
```suggestion
class AspectRatio(Enum):
```

File: pyface/ui/qt4/util/image_helpers.py
Comment (rahulporuri) [inline pyface/ui/qt4/util/image_helpers.py:38]: i noticed some white space here

```suggestion

```

File: pyface/ui/qt4/util/image_helpers.py
Comment (rahulporuri) [inline pyface/ui/qt4/util/image_helpers.py:54]: white space here too 

```suggestion

```

File: pyface/ui/qt4/util/image_helpers.py
Comment (rahulporuri) [inline pyface/ui/qt4/util/image_helpers.py:70]:
```suggestion

```

File: pyface/ui/qt4/util/image_helpers.py
Comment (rahulporuri) [inline pyface/ui/qt4/util/image_helpers.py:178]: do want to remove these in the next major release? i'm not sure if anyone outside of pyface uses them directly

File: pyface/ui/wx/util/image_helpers.py
Comment (rahulporuri) [inline pyface/ui/wx/util/image_helpers.py:101]:
```suggestion
    This copies the data returned from wx.
```

File: pyface/ui/wx/util/image_helpers.py
Comment (rahulporuri) [inline pyface/ui/wx/util/image_helpers.py:131]:
```suggestion
    This copies the data before passing it to wx.
```

File: pyface/ui/wx/util/image_helpers.py
Comment (rahulporuri) [inline pyface/ui/wx/util/image_helpers.py:143]: it looks like this function might return an image with 4 channels so shouldn't this be `RGB32 or ARGB32`?

File: pyface/ui_traits.py
Comment (rahulporuri) [inline pyface/ui_traits.py:45]:
```suggestion
# cache of conversions of ImageResource instances to toolkit bitmaps
```

File: pyface/util/image_helpers.py
Comment (rahulporuri) [inline pyface/util/image_helpers.py:16]:
```suggestion
This module provides helper functions for converting between numpy arrays
and toolkit images, as well as between the various image types in a standardized
way.

Helper functions
----------------

- :data:`~.array_to_image`
- :data:`~.bitmap_to_icon`
- :data:`~.bitmap_to_image`
- :data:`~.image_to_array`
- :data:`~.image_to_bitmap`
- :data:`~.resize_image`
- :data:`~.resize_bitmap`

Options for resizing images
---------------------------
- :data:`~.ScaleMode`
- :data:`~.AspectRatio`

"""
```

**Review by rahulporuri — APPROVED** [review]
Mostly LGTM with a bunch of comments/corrections/suggestions.

File: pyface/array_image.py
Comment (corranwebster) [reply pyface/array_image.py:1]: The `pyface.image` subpackage is all about loading images from zipfile libraries of images.  These modules are at the same level or above `pyface.image_resources`, and so should live at the same level.

File: pyface/i_image_resource.py
Comment (corranwebster) [reply pyface/i_image_resource.py:67]: No - this came from Wx wanting `wx.Bitmap` objects in many cases, where Qt in the same situation likes `QImage` objects (or sometimes vice-versa).  Part of this change is making the distinction more concrete, but toolkits won't change what they actually use.

File: pyface/ui/qt4/util/image_helpers.py
Comment (corranwebster) [reply pyface/ui/qt4/util/image_helpers.py:178]: They might be being used in enable, but yes, they should probably go in the next major release.

File: pyface/ui/wx/util/image_helpers.py
Comment (corranwebster) [reply pyface/ui/wx/util/image_helpers.py:143]: Actually this is a search and replace bug.  `wx.Image` doesn't have a notion of format, it's all "RGB" with an optional seperate alpha channel:
```suggestion
        The wx.Image created from the data.
```

File: pyface/util/image_helpers.py
Comment (corranwebster) [reply pyface/util/image_helpers.py:16]: I'm assuming this is needed because of the way that API docs are generated?

File: pyface/ui/qt4/util/image_helpers.py
Comment (corranwebster) [inline pyface/ui/qt4/util/image_helpers.py:176]:
```suggestion
# backwards compatible names - will be removed in Pyface 8
```

**Comment by rahulporuri** [comment]
@corranwebster can you also merge the master branch into this branch? It looks like GitHub Actions isnt running on this branch. Or you could potentially wait until after #967 is merged.

File: pyface/ui/qt4/util/image_helpers.py
Comment (corranwebster) [reply pyface/ui/qt4/util/image_helpers.py:178]: I've added a not to this effect, and will open an issue.

File: pyface/util/image_helpers.py
Comment (rahulporuri) [reply pyface/util/image_helpers.py]: yep. i thought it'll make the autogenerated api docs better.

File: pyface/ui/qt4/util/image_helpers.py
Comment (rahulporuri) [reply pyface/ui/qt4/util/image_helpers.py:175]: thank you!

Comment (rahulporuri) [reply pyface/ui/qt4/util/image_helpers.py:175]: and it doesn't look like they're used outside pyface - https://github.com/search?q=org%3Aenthought+QImage_to_array&type=code and https://github.com/search?q=org%3Aenthought+array_to_QImage&type=code

**Comment by corranwebster** [comment]
The changes to the Enums broke something - looking into it.

**Comment by corranwebster** [comment]
OK, looks like they might need to be `IntEnum`; or we shouldn't use enums at all (this was an experiment on my part).

**Pull request merged**
